File: games/autoclicker.js
const robot = require('robotjs');
const readlineSync = require('readline-sync');

let clicking = false;
let clickIntervalId = null; // Untuk menyimpan ID dari setInterval

function performClick() {
    // Ambil posisi mouse saat ini dan klik
    const mousePos = robot.getMousePos();
    console.log(`Melakukan klik pada: ${mousePos.x}, ${mousePos.y}`);
    robot.mouseClick();
}

function startClicking(delayMs, numClicksStr) {
    if (clicking) {
        console.log("Klik otomatis sudah berjalan.");
        return;
    }

    let numClicks;
    if (numClicksStr.toLowerCase() === 'inf' || numClicksStr.toLowerCase() === 'infinite') {
        numClicks = Infinity;
    } else {
        numClicks = parseInt(numClicksStr);
        if (isNaN(numClicks) || numClicks <= 0) {
            console.log("Jumlah klik tidak valid. Harus angka positif atau 'inf'.");
            return;
        }
    }

    if (isNaN(delayMs) || delayMs <= 0) {
        console.log("Jeda tidak valid. Harus angka positif.");
        return;
    }

    clicking = true;
    let clicksDone = 0;
    console.log(`Memulai klik otomatis dengan jeda ${delayMs} ms.`);
    if (numClicks === Infinity) {
        console.log("Klik akan berjalan tanpa batas. Tekan CTRL+C untuk berhenti.");
    } else {
        console.log(`Akan melakukan ${numClicks} klik. Tekan CTRL+C untuk berhenti lebih awal.`);
    }

    clickIntervalId = setInterval(() => {
        if (!clicking) {
            clearInterval(clickIntervalId);
            return;
        }

        performClick();
        clicksDone++;

        if (clicksDone >= numClicks) {
            stopClicking();
            console.log(`${clicksDone} klik telah dilakukan. Selesai.`);
        }
    }, delayMs);
}

function stopClicking() {
    if (clicking) {
        clearInterval(clickIntervalId);
        clicking = false;
        console.log("Klik otomatis dihentikan.");
    }
}

function main() {
    console.log("--- Auto Clicker Sederhana (Node.js) ---");
    console.log("PENTING: Pindahkan kursor mouse ke posisi yang diinginkan sebelum klik dimulai.");
    console.log("Tekan CTRL+C di terminal ini untuk keluar atau menghentikan paksa kapan saja.\n");

    const delayInput = readlineSync.question("Masukkan jeda antar klik (dalam milidetik, contoh: 100): ");
    const numClicksInput = readlineSync.question("Masukkan jumlah klik (angka, atau 'inf' untuk tak terbatas): ");

    const delayMs = parseInt(delayInput);

    // Beri waktu pengguna untuk memindahkan mouse setelah input
    console.log("\nAnda memiliki 5 detik untuk memindahkan kursor mouse ke posisi target...");
    setTimeout(() => {
        console.log("Memulai...");
        startClicking(delayMs, numClicksInput);
    }, 5000); // Jeda 5 detik

    // Menjaga skrip tetap berjalan sampai dihentikan manual jika klik tak terbatas
    // atau sampai selesai jika jumlah klik terbatas.
    // CTRL+C akan menghentikan proses Node.js secara keseluruhan.
    process.on('SIGINT', () => {
        console.log("\nCTRL+C ditekan.");
        stopClicking();
        process.exit();
    });
}

main();

File: games/autoclicker.py
import time
import threading
import mouse # Untuk kontrol mouse
import keyboard # Untuk kontrol keyboard

# Variabel global untuk mengontrol status klik
clicking = False
stop_event = threading.Event()
click_thread = None

def perform_click():
    """Melakukan satu kali klik mouse kiri pada posisi kursor saat ini."""
    mouse.click('left')

def auto_clicker(delay_ms, num_clicks_str):
    """Fungsi utama untuk melakukan klik otomatis."""
    global clicking
    clicking = True
    print(f"Memulai klik otomatis dengan jeda {delay_ms} ms.")

    if num_clicks_str.lower() == 'inf':
        num_clicks = float('inf')
        print("Klik akan berjalan tanpa batas hingga dihentikan (Tekan 's' lagi).")
    else:
        try:
            num_clicks = int(num_clicks_str)
            if num_clicks <= 0:
                print("Jumlah klik harus lebih dari 0.")
                clicking = False
                return
            print(f"Akan melakukan {num_clicks} klik.")
        except ValueError:
            print("Jumlah klik tidak valid. Masukkan angka atau 'inf'.")
            clicking = False
            return

    clicks_done = 0
    while clicking and clicks_done < num_clicks:
        if stop_event.is_set():
            break
        perform_click()
        clicks_done += 1
        if clicks_done < num_clicks: # Hanya tidur jika masih ada klik selanjutnya
            time.sleep(delay_ms / 1000.0) # Konversi milidetik ke detik
        if clicks_done % 10 == 0 and num_clicks == float('inf'):
             print(f"{clicks_done} klik telah dilakukan...")


    if not stop_event.is_set(): # Hanya cetak jika tidak dihentikan secara paksa
        print("Klik otomatis selesai.")
    clicking = False
    stop_event.clear()

def toggle_clicking():
    """Memulai atau menghentikan proses klik otomatis."""
    global clicking, click_thread, stop_event

    if clicking:
        print("Menghentikan klik otomatis...")
        stop_event.set()
        if click_thread and click_thread.is_alive():
            click_thread.join()
        clicking = False
        print("Klik otomatis dihentikan.")
    else:
        if click_thread and click_thread.is_alive():
            print("Proses klik sebelumnya masih menyelesaikan, harap tunggu.")
            return

        try:
            delay_input = input("Masukkan jeda antar klik (dalam milidetik): ")
            delay_ms = int(delay_input)
            if delay_ms <= 0:
                print("Jeda harus lebih dari 0 milidetik.")
                return

            num_clicks_input = input("Masukkan jumlah klik (angka, atau 'inf' untuk tak terbatas): ")

            stop_event.clear()
            click_thread = threading.Thread(target=auto_clicker, args=(delay_ms, num_clicks_input))
            click_thread.start()

        except ValueError:
            print("Input jeda tidak valid. Harap masukkan angka.")
        except Exception as e:
            print(f"Terjadi kesalahan: {e}")

def main():
    """Fungsi utama untuk menjalankan program."""
    print("--- Auto Clicker Sederhana (Alternatif) ---")
    print("Tekan 's' untuk memulai/menghentikan klik otomatis.")
    print("Tekan 'q' untuk keluar dari program.")
    print("-------------------------------------------")
    print("PENTING: Pastikan kursor mouse berada pada posisi yang diinginkan sebelum menekan 's'.")
    print("CATATAN: Program ini mungkin memerlukan hak akses administrator (root) untuk fungsi keyboard.")

    # Menambahkan hotkey
    # `suppress=True` mencegah event tombol 's' diketik di aplikasi lain saat program berjalan
    # Namun, ini bisa mengganggu jika pengguna benar-benar ingin mengetik 's'
    # Jadi, kita bisa menghapusnya atau menggantinya dengan `keyboard.on_press_key("s", lambda _: toggle_clicking())`
    # yang memungkinkan event terus berlanjut. Untuk kasus ini, tidak men-suppress mungkin lebih baik.

    keyboard.add_hotkey('s', toggle_clicking)
    keyboard.add_hotkey('S', toggle_clicking) # Untuk Shift + s

    print("Listener keyboard aktif. Tekan 's' atau 'q'.")

    # Menjaga program tetap berjalan hingga 'q' ditekan
    # `keyboard.wait('q')` akan memblokir hingga 'q' ditekan
    # dan kemudian menghentikan semua hotkey yang terdaftar.
    try:
        keyboard.wait('q', suppress=True) # Menunggu tombol 'q' untuk keluar
        print("\nTombol 'q' ditekan. Keluar dari program...")
    except Exception as e:
        print(f"Error saat menunggu tombol keluar: {e}")
    finally:
        global clicking, stop_event, click_thread
        if clicking: # Jika sedang mengklik saat 'q' ditekan
            stop_event.set()
            if click_thread and click_thread.is_alive():
                click_thread.join()
        keyboard.remove_all_hotkeys() # Membersihkan semua hotkey


if __name__ == "__main__":
    main()